### PR TITLE
Top bar

### PR DESCRIPTION
**Pull request by alrusdi**
Optional sticky top bar. I don't like visuals, but it is probably ok as a start...
sticky

Demo: http://37.230.117.156:9000/

### CODE REVIEW

**Comment by kberg** [comment]
I'll look tomorrow evening. It's a great start.

File: src/components/Preferences.ts [hunk 1]
```diff
@@ -247,6 +248,12 @@ export const Preferences = Vue.component('preferences', {
                             <i class="form-icon"></i> <span v-i18n>Hide corporation names for players</span>
                         </label>
                     </div>
+                    <div class="preferences_panel_item">
```
Comment (kberg) [inline src/components/Preferences.ts:251]: Instead, I'd make it dismissable and redisplayable right on the bar. You've seen this type of thing, of course. And that toggle can be stored as the preference.

Comment (kberg) [reply src/components/Preferences.ts:251]: ... or both, now that I think about it.

**Review by kberg — COMMENTED** [review]
Basically LG, but someone else should opine about CSS and such.

**Review by chosta — COMMENTED** [review]
Thanks for making that optional!
I have no issues with vue and logic. Lets see how the visuals will turn out to be and let me know if you need help.

**Comment by alrusdi** [comment]
> Thanks for making that optional!
> I have no issues with vue and logic. Lets see how the visuals will turn out to be and let me know if you need help.

If we will make decision to make top bar removable in place (as @kberg suggested above) then i need some visual solution for it. Just "close button" on the right side of the bar looks odd

**Comment by chosta** [comment]
How about an indicator that slides the bar left / right? Example - Evernote web interface:

sidebar-1

File: src/components/Preferences.ts (hunk 1 above)
Comment (alrusdi) [reply src/components/Preferences.ts:251]: > ... or both, now that I think about it.

I've implemented both

File: src/components/TopBar.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  },
+  data: function() {
+    return {};
+  },
+  methods: {
+    toggleBar() {
+      const newVal = this.isExpanded() ? '1': '';
+      PreferencesManager.saveValue('hide_top_bar', newVal);
+      PreferencesManager.preferencesValues.set('hide_top_bar', this.isExpanded());
+      this.$forceUpdate();
```
Comment (bafolts) [inline src/components/TopBar.ts:23]: You can avoid using `$forceUpdate` if a data property is created for `hideTopBar`. When that property were to change the template would update without needing to call to force the update. The `hideTopBar` would default to `PreferencesManager.loadValue('hide_top_bar')`.

**Review by bafolts — CHANGES_REQUESTED** [review]
I would avoid using `$forceUpdate` and use data properties to let vue handle updating the component naturally without forcing it.

File: src/components/TopBar.ts (hunk 2 above)
Comment (alrusdi) [reply src/components/TopBar.ts]: Your review lead me to read this article https://michaelnthiessen.com/force-re-render/
Author's point seems to be reasonable, so I've used his technique to force the component update.

File: src/components/TopBar.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    PlayerInfo,
+  },
+  data: function() {
+    return {
+      componentKey: 0,
+    };
+  },
+  methods: {
+    forceRerender() {
+      this.componentKey += 1;
```
Comment (bafolts) [inline src/components/TopBar.ts:22]: This is at least not using $forceUpdate but still seems to be working against vue. With vue we shouldn't have to be concerned with when the renderer is going to update and it should more or less happen naturally as the data is updated.  If this used:

```
data: function() {
  return {
    hideTopBar: PreferencesManager.preferencesValue.loadValue('hide_top_bar') !== '1'
  }
},
methods: {
   toggleBar(): {
     this.componentKey = this.componentKey === '1' ? '' : '1'
     PreferencesManager.saveValue('hide_top_bar', this.componentKey);
  }
}
```
The template will be updated as we are naturally updating `hideTopBar`. Anyone updating `hideTopBar` then doesn't need to call to `forceRenderer`. As this is at least not using $forceUpdate though it is good enough.

**Review by bafolts — APPROVED** [review]

**Pull request merged**